resolve.go

```diff
@@ -44,25 +45,38 @@ func (h *handler) ServeDNS(w dns.ResponseWriter, r *dns.Msg) {
 }
 
 func (h *handler) resolveDnsNames(names []string) (*dns.Msg, error) {
-	h.log.Info("forwarding request", "resolver", inputConfig.dnsServer)
 
 	dnsClient := h.dnsClient
+	dnsServer := h.dnsServer
 
 	for _, name := range names {
+		h.log.Info("forwarding request", "resolver", dnsServer)
 		attempt := 1
+		resolveDnsServerOnce := false
 
 	Redo:
 		dnsRequest := new(dns.Msg)
 		dnsRequest.SetQuestion(name, dns.TypeA)
 		dnsRequest.SetEdns0(4096, true)
 
 		h.log.Info("doing dns lookup", "req", dnsRequest)
-		ans, rtt, err := dnsClient.Exchange(dnsRequest, inputConfig.dnsServer)
+		ans, rtt, err := dnsClient.Exchange(dnsRequest, dnsServer)
 		if err != nil {
+			// if it is a dial error, retry after resolving dns server again
+			if _, ok := err.(*net.OpError); ok && !resolveDnsServerOnce {
+				h.log.Info(fmt.Sprintf("Received Dial Error: %v. Resolved DNS Server Again and Retrying dns lookup. Attempt: %v", err, attempt))
+				dnsServer = resolveDnsServerAddress(h.log)
+				resolveDnsServerOnce = true
+				goto Redo
+			}
```

Comment:
How do we handle temporary errors? I think this will permanently fallback to other DNS server. What we probably need is:
1. Have two (or more) DNS resolvers: example, first one being kube-dns and second one is VPC etc.
2. If first lookup fails, fallback to second one?

Reply:
No @surki. For that request, it will fallback to the VPC one even after resolving again produces dial error. For subsequent requests, it will make use of the kube-dns only.

Reply:
Okay, looking at `resolveDnsServerAddress`, it seems we are making kube api call in the failure path. The kube-dns address is not going to change, why not lookup that in startup and then use it forever?

Reply:
We are doing during instantiation also Suresh. https://github.com/surki/dns-zone-aware/pull/2/files#diff-cb1b8a01bbc5f64cf06dce0e3f8021ad8289078b402dc620afa04780484d7cd5R98 Redoing it only once if failure is dial error.